In [1]:
pip install -q hvplot

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install -q holoviews

Note: you may need to restart the kernel to use updated packages.


In [3]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [8]:


# Read the CSV file into a DataFrame
credit_risk_df = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
credit_risk_df.head()


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [10]:
# Separate the data into labels and features

# Separate the Y variable, the labels
Y= credit_risk_df['loan_status']

# Separate the X variable, the labels
X= credit_risk_df.drop(columns = 'loan_status')


In [11]:
# Review the y variable Series
Y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [12]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [14]:
# Check the balance of our target values
Y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [15]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [16]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
LRM = LogisticRegression( random_state=1)

# Fit the model using training data
LRM.fit(X_train, Y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [17]:
# Make a prediction using the testing data
PredictX= LRM.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [20]:
# Print the balanced_accuracy score of the model

from sklearn.metrics import accuracy_score

AS = accuracy_score(Y_test, PredictX)
print("Accuracy Score:", AS)


Accuracy Score: 0.9918489475856377


In [26]:
#Round the accuracy score
AS1 = round(AS*100,2)
AS1
print("Accuracy Score:", AS1)


Accuracy Score: 99.18


In [29]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix
confusionMatrix = confusion_matrix(Y_test, PredictX)
print(confusionMatrix)

[[18663   102]
 [   56   563]]


In [31]:
# Print the classification report for the model
from sklearn.metrics import classification_report

report = classification_report(Y_test, PredictX, target_names=["Healthy", "High Risk"])
print(report)


              precision    recall  f1-score   support

     Healthy       1.00      0.99      1.00     18765
   High Risk       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistric regression model has a high accuracy score of 99.18%.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [33]:
pip install imbalanced-learn


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 226.0/226.0 kB 2.1 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 823.6 kB/s eta 0:00:0000:0100:01
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.0
    Uninstalling joblib-1.1.0:
      Successfully uninstalled joblib-1.1.0
Note: you may need to restart the kernel to use updated packages.


In [34]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
RandomSampler= RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, Y_resampled = RandomSampler.fit_resample(X_train, Y_train)

In [36]:
# Count the distinct values of the resampled labels data
Y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [38]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
RSM2= LogisticRegression(random_state =1)
# Fit the model using the resampled training data
RSM2.fit(X_resampled, Y_resampled)

# Make a prediction using the testing data
Predict2 = RSM2.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [41]:
# Print the balanced_accuracy score of the model 
AS2 = accuracy_score(Y_test, Predict2)
rounded_AS2 = round(AS2*100, 2)
print("Accuracy Score:", rounded_AS2)


Accuracy Score: 99.38


In [43]:
# Generate a confusion matrix for the model
confusion2 = confusion_matrix(Y_test, Predict2)
print(confusion2)

[[18649   116]
 [    4   615]]


In [45]:
from sklearn.metrics import classification_report

report2 = classification_report(Y_test, Predict2, target_names=["Healthy", "High Risk"])
print(report2)


              precision    recall  f1-score   support

     Healthy       1.00      0.99      1.00     18765
   High Risk       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The accuracy score for the oversampled data (99.38) is higher than the accuracy score for the original data (99.18). 